tycho-extras/tycho-pomless/src/main/java/org/eclipse/tycho/pomless/AbstractTychoMapping.java

```diff
@@ -261,20 +258,47 @@ protected Charset getPrimaryArtifactCharset() {
         return StandardCharsets.UTF_8;
     }
 
-    protected abstract void initModel(Model model, Reader artifactReader, File artifactFile)
-            throws ModelParseException, IOException;
+    protected abstract void initModel(Model model, Reader artifactReader, File artifactFile) throws IOException;
 
     protected static Properties getBuildProperties(File dir) throws IOException {
+        return loadProperties(new File(dir, "build.properties"));
+    }
+
+    static Properties loadProperties(File propertiesPath) throws IOException {
         Properties properties = new Properties();
-        File buildProperties = new File(dir, "build.properties");
-        if (buildProperties.exists()) {
-            try (FileInputStream stream = new FileInputStream(buildProperties)) {
+        if (propertiesPath.isFile()) {
+            try (FileInputStream stream = new FileInputStream(propertiesPath)) {
                 properties.load(stream);
             }
         }
         return properties;
     }
 
+    static Supplier<Properties> getPropertiesSupplier(File propertiesPath) {
+        FutureTask<Properties> t = new FutureTask<>(() -> loadProperties(propertiesPath));
+        return () -> {
+            try { // Load properties on first demand and cache result
+                t.run(); // No-Op if already running
+                return t.get();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            } catch (ExecutionException e1) { // ignore externalzied data
+            }
```

Comment:
I just wondered if a `IllegalStateException` should be thrown here using the `ExecutionException`'s cause as its cause?
Before the `TychoBundleMapping` propagates exceptions while loading the localization while the `TychoFeatureMapping` ignored the externalized data when loading the properties failed with an exception.

I think it would be better to propagate the exception to not swallow the failure.

Reply:
Localization is something one could assume as optional and thus I don't think a failure is appropriate.

Reply:
I'm not sure if this is such a minor problem then.
For example many Eclipse Plugins use `%Bundle-Name` and `%Bundle-Vendor` in the Manifest headers for their name/vendor. Imagine the read fails on the release build (for whatever reason and for whatever reason it only fails for that file), then we publish a Plug-in that uses that variable name as its Bundle-Name/Vendor.
It probably won't break anything, but at least it is nasty. And as somebody who builds and publishes some bundles I would like to get an error.

Actually I would even go one step further and fail if the substitution for the variable is not present in order to detect errors where a variable is used but not specified.

Reply:
This is only used here to fill the maven info, the goal of the pomless build is not to validate things, this could be done in the actual build if required. Also internationalization could be part of a fragment as well.

Reply:
> This is only used here to fill the maven info, the goal of the pomless build is not to validate things, this could be done in the actual build if required. Also internationalization could be part of a fragment as well.

Indeed. I didn't consider that. Then we can treat this as optional.
But then localizedValue() should return null in case a translation is not found to not write `%whatever` into the pom?

Reply:
The translation is applied as "best effort", in general the contract for pomless is that it tries to extract as much information as possible, if you follow the convention, if not one has to customize it (e.g. build.properties) or use `pom.xml`

Reply:
> The translation is applied as "best effort", in general the contract for pomless is that it tries to extract as much information as possible, if you follow the convention, if not one has to customize it (e.g. build.properties) or use `pom.xml`

Yes, agree as well. But do you think a value to translate (i.e. starting with a '%') should be kept if no translation is found?
In this case I think it would be better to assume no such value exists at all and stick with the default, since with "best effort" no actual value is found.

Reply:
This should already be the case that the value is retained but I'm not sure if maybe the code behaves like if emtpy value then no name is set at all...

Reply:
At the moment `TychoBundleMapping` and `TychoFeatureMapping` simply uses the 'raw-value' of a property if it starts with `%` and there is no translation available.
It would IMHO be better to assume no value than, than using a value that actually has to be translated first.

Reply:
> At the moment `TychoBundleMapping` and `TychoFeatureMapping` simply uses the 'raw-value' of a property if it starts with `%` and there is no translation available.

That's how it is meant to be, the "key" is not a technical key but a "default value if no translation is available" see http://docs.osgi.org/specification/osgi.core/7.0.0/framework.module.html#i3189742

Reply:
OK. I changed it to return the 'key' (i.e. the raw-value without %-prefix) if no translation is found.
After checking again this was the behavior for bundles but not for features (which just used the raw-value as it was).

Reply:
Yes it might be inconsistent (aka not tested) here for features :-)